### PR TITLE
feat: spread load of rule evaluation by group when starting new groups

### PR DESCRIPTION
**Pull request by dxtrzhang**
spread load of rule evaluation by group

### CODE REVIEW

**Comment by dxtrzhang** [comment]
FYI. This PR may make rule file reloading time even worse, see #691 and #704.
Pls resolve that PR first.

File: app/vmalert/group.go [hunk 1]
```diff
@@ -182,10 +184,22 @@ func (g *Group) close() {
...
+	randSleep := uint64(float64(g.Interval) * (float64(h) / (1 << 32)))
+	sleeper := time.NewTimer(time.Duration(randSleep))
+	var t *time.Ticker
+	var e *executor
+	select {
+	case <-g.finishedCh:
+		sleeper.Stop()
+		return
+	case <-sleeper.C:
+		logger.Infof("sleep for %v, group %q started; interval=%v; concurrency=%d", time.Duration(randSleep), g.Name, g.Interval, g.Concurrency)
```
Comment (valyala) [inline app/vmalert/group.go:198]: It would be better moving lines 198-201 below the `select` block. This should reduce the resulting diff.

**Review by valyala — COMMENTED** [review]
LGTM overall. Please address the comment in order to reduce the resulting diff for this pull request.

**Review by valyala — APPROVED** [review]
LGTM.

@hagen1778 , could you look at the change and merge if it is OK?

File: app/vmalert/group.go [hunk 2]
```diff
@@ -182,7 +184,17 @@ func (g *Group) close() {
...
+	h := uint32(xxhash.Sum64([]byte(strconv.FormatUint(g.ID(), 10))))
+	randSleep := uint64(float64(g.Interval) * (float64(h) / (1 << 32)))
+	sleeper := time.NewTimer(time.Duration(randSleep))
+	select {
+	case <-g.finishedCh:
+		sleeper.Stop()
+		return
+	case <-sleeper.C:
+	}
+	logger.Infof("sleep for %v, group %q started; interval=%v; concurrency=%d", time.Duration(randSleep), g.Name, g.Interval, g.Concurrency)
```
Comment (hagen1778) [inline app/vmalert/group.go:197]:
```suggestion
	logger.Infof("group %q started with delay %v; interval=%v; concurrency=%d", time.Duration(randSleep), g.Name, g.Interval, g.Concurrency)
```

**Review by hagen1778 — CHANGES_REQUESTED** [review]
Looks good! See my comments

File: app/vmalert/group.go [hunk 3]
```diff
@@ -182,7 +184,17 @@ func (g *Group) close() {
 
 func (g *Group) start(ctx context.Context, querier datasource.Querier, nts []notifier.Notifier, rw *remotewrite.Client) {
 	defer func() { close(g.finishedCh) }()
-	logger.Infof("group %q started; interval=%v; concurrency=%d", g.Name, g.Interval, g.Concurrency)
+	// This should spread load of rule evaluation by group
```
[the remark this answers is not in the document]
Comment (valyala) [reply app/vmalert/group.go:187]: Addressed these comments in 5f16ceb294946a48a65e958fa8acd91a2e266ad9

**Comment by valyala** [comment]
@dxtrzhang , thanks for the contribution! Could you verify whether `vmalert` built from the commit 5f16ceb294946a48a65e958fa8acd91a2e266ad9 in `master` branch spreads group evaluation load over time as expected?

**Comment by dxtrzhang** [comment]
works as expected！
Thanks for the update :)

**Comment by valyala** [comment]
FYI, the feature has been included in [v1.40.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

**Pull request merged**